executor/src/witgen/jit/solver/mod.rs

```diff
@@ -0,0 +1,212 @@
+            variable_states,
+            input_variables,
+        }
+    }
+
+    /// Solves the constraints as far as possible. For each variable it was able to
+    /// find a unique solution for, the result contains an expression, indicating
+    /// how the variable can be computed from other variables at runtime.
+    /// If we were able to solve for the value already, the expression will be an
+    /// instance of `SymbolicExpression::Concrete(_)`.
+    #[allow(dead_code)]
+    pub fn solve(mut self) -> Result<BTreeMap<V, SymbolicExpression<T, V>>, Error> {
+        self.loop_until_no_progress()?;
+
+        // Return assignment for all known variables
+        Ok(self
+            .variable_states
+            .into_iter()
+            .filter_map(|(v, info)| info.symbolic_expression.map(|expr| (v, expr)))
+            .filter(|(v, _)| !self.input_variables.contains(v))
+            .collect())
+    }
+
+    fn loop_until_no_progress(&mut self) -> Result<(), Error> {
+        loop {
+            let mut progress = false;
+            for i in 0..self.algebraic_constraints.len() {
+                // TODO: Improve efficiency by only running skipping constraints that
+                // have not received any updates since they were last processed.
+                let effects = self.algebraic_constraints[i].solve(self)?.effects;
+                for effect in effects {
+                    progress |= self.apply_effect(effect);
+                }
+            }
+            if !progress {
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    fn apply_effect(&mut self, effect: Effect<T, V>) -> bool {
+        match effect {
+            Effect::Assignment(v, expr) => self.apply_assignment(v, expr),
+            Effect::RangeConstraint(v, range_constraint) => {
+                self.apply_range_constraint_update(v, range_constraint)
+            }
+            // TODO: We need to introduce a new operator in `SymbolicExpression` for this.
+            Effect::BitDecomposition(..) => unimplemented!(),
+            // QuadraticSymbolicExpression::solve() never returns these
+            Effect::MachineCall(..)
+            | Effect::Assertion(..)
+            | Effect::ProverFunctionCall(..)
+            | Effect::Branch(..) => unreachable!(),
+        }
+    }
+
+    fn apply_assignment(&mut self, variable: V, expr: SymbolicExpression<T, V>) -> bool {
+        // If the expression's range constraint tighter than the current one, we need to
+        // update it.
+        let progress_range_constraint =
+            self.apply_range_constraint_update(variable.clone(), expr.range_constraint());
+
+        let entry = self.variable_states.entry(variable.clone()).or_default();
+        let variable_update = match &entry.symbolic_expression {
+            // We already know the expression
+            Some(existing) if existing == &expr => None,
+
+            // The expression was updated. This should never happen.
```

Comment:
Yes it should. Example:
```
Y = Z;
X = Y;
Y = 8;
```
`Z` is known, we process in order. In the end I think we want to know that all of them are `8`. This is especially useful when one of them is a guard: `(X - 8) * (T - 19) = 0`